### ISSUE TITLE
Server acknowledging index creation, but it doesn't really happen

### ISSUE THREAD

**Issue by marshall007** [rethinkdb/rethinkdb#4694]
When I try to create multiple indices in parallel the server acknowledges the creation of each one, but only one is actually created. Here's a simple script to reproduce (I'm on v2.1.1):

```js
var r = require('rethinkdb');
var Promise = require('bluebird');

r.connect({ db: 'test' })
.then(function(conn) {
  return Promise.map(['foo', 'bar', 'baz'], function(index) {
    console.log('creating:', index);
    return r.table('test').indexCreate(index).run(conn);
  })
  .then(function(data) {
    console.log(data);
    return r.table('test').indexList().run(conn);
  }).then(function(data) {
    return console.log(data);
  })
})
```

```
creating: foo
creating: bar
creating: baz
[ { created: 1 }, { created: 1 }, { created: 1 } ]
[ 'baz' ]
```

**Comment by danielmewes** [comment]
Thanks for the report and test code @marshall007 . I can reproduce this, and I think this is because we apply the index change to the configuration non-atomically.

**Comment by danielmewes** [comment]
@VeXocide This is actually slightly tricky, and I'd like to hear your input on it.

If you look at `real_reql_cluster_interface_t::sindex_change_internal()`, you can see that we get the table config, then apply the change, and then set the new config.

The other place where this might become a problem is when writing to the `table_config` document (let's say you use an `update` query to update one field of the table configuration at a time).

Do you think it would be reasonable to add a version or UUID to `table_config_and_shards_t`, so that we could detect if the configuration had changed since we last read it?
We would send the version we're expecting to the Raft leader when setting the new config, and it would return an error if the config no longer matches the current version.

**Comment by danielmewes** [comment]
(actually we could probably use the `multi_table_manager_timestamp_t` for that?)

**Comment by VeXocide** [comment]
This is next, after talking with @danielmewes we're positive that we've found a solution.

He mentioned above that changing the secondary indexes consists of fetching the committed `table_config_and_shards_t`, applying the change, and subsequently setting the new config by calling `contract_coordinator_t::change_config` which in turn proposes it to raft. It may take some time for this proposed raft change to be committed, which is why you end up with only the last index.

The solution is twofold, first of all we don't want to manipulate the committed state, but the latest state which is already possible using `contract_coordinator_t::change_config`. Secondly we don't want to change the whole configuration but only a single aspect, changing the table name or adding an index for example.

**Comment by VeXocide** [comment]
This is in CR 3178 by @danielmewes.

**Comment by VeXocide** [comment]
Merged into branches `next` and `v2.1.x` via commits 2cfe68136e618110323c024b1df5e8d3b0f3d186 and 6bb34b84c883c18425cbf991e4882f59cedac8bd respectively.

**Comment by danielmewes** [comment]
@marshall007 We released 2.1.2 with a fix for this.

**Comment by marshall007** [comment]
@danielmewes thanks!
